pb/state/state.proto

```diff
@@ -40,21 +136,128 @@ message Row {
 }
 
+// A single table of test results backing a dashboard tab.
 message Grid {
+  // A cycle of test results, not including the results. In the TestGrid client,
+  // the cycles define the columns.
   repeated Column columns = 1;
+
+  // A test case with test results. In the TestGrid client, the cases define the
+  // rows (and the results define the individual cells).
   repeated Row rows = 2;
 
-  reserved 3 to 9;  // Consider later
+  // Seconds since epoch when last mail alert was sent.
+  double last_alert_mail_time = 3;
+
+  // The latest configuration used to generate this test group.
+  TestGroup config = 4;
+
+  reserved 5;
+
+  // Seconds since epoch for last time this cycle was updated.
+  double last_time_updated = 6;
+
+  reserved 7;
+
+  // Stored info on previous timing for parts of the update cycle.
+  // TODO(b/63145995): Move last_time_updated, last_alert_mail_time here.
+  repeated UpdateInfo update_info = 8;
+
+  // Stored info on default test metadata.
+  repeated TestMetadata test_metadata = 9;
+
+  // Clusters of failures for a TestResultTable instance.
+  repeated Cluster cluster = 10;
+
+  // Most recent timestamp that clusters have processed.
+  double most_recent_cluster_timestamp = 11;
+}
+
+// A cluster of failures grouped by test status and message for a test results
+// table.
+message Cluster {
+  // Test status cluster grouped by.
+  int32 test_status = 1;
+
+  // Error message or testFailureClassification string cluster grouped by.
+  string message = 2;
+
+  // ClusterRows that belong to this cluster.
+  repeated ClusterRow cluster_row = 3;
 }
+
+// Cells in a TestRow that belong to a specific Cluster.
+message ClusterRow {
+  // Name of TestRow.
+  string display_name = 1;
+
+  // Index within row that belongs to Cluster (refer to columns of the row).
+  repeated int32 index = 2;
+}
+
+// An identifier of a dashboard tab, i.e., the name of the dashboard and tab.
+message DashboardTabIdentifier {
+  // The name of a dashboard containing the dashboard tab.
+  string dashboard_name = 1;
+
+  // The dashboard tab to update.
+  DashboardTab dashboard_tab = 2;
+}
+
+// An updater request to update a test group.
+message UpdateRequest {
+  // The test group configuration to update.
+  TestGroup test_group = 1;
+
+  // A list of dashboards tabs backed by the group being updated.
+  repeated DashboardTabIdentifier dashboard_tab_identifiers = 2;
+}
+
+// An updater response after updating a test group.
+message UpdateResponse {
+  // The time taken to perform the update in milliseconds.
+  uint32 update_time_millis = 1;
+
+  // The size of the compressed output file in bytes.
+  uint32 output_size_bytes = 2;
+
+  // Summaries of dashboard tabs associated with this group.
+  repeated DashboardTabSummary dashboard_tab_summaries = 3;
+
+  // Metrics associated with the update for this group.
+  UpdateInfo update_entry = 4;
+}
+
+// An updater server updates test groups upon request.
+service Updater {
```

Comment:
I suspect this is going to import a lot of unnecessary junk. Thoughts on moving update RPCs into their own proto?

Reply:
That's reasonable yeah, let me go do that.

Reply:
Done!